File: Besties/Listas/lista_circular.cpp
#include "lista_circular.h"
#include <iostream>

ListaCircular::ListaCircular() : ListaBase(true) {}

void ListaCircular::insertar(const Persona& p) {
    Nodo* nuevoNodo = new Nodo(p);
    
    if (!cabeza) {
        cabeza = nuevoNodo;
        cabeza->setSiguiente(cabeza);
    } else {
        Nodo* temp = cabeza;
        while (temp->getSiguiente() != cabeza) {
            temp = temp->getSiguiente();
        }
        temp->setSiguiente(nuevoNodo);
        nuevoNodo->setSiguiente(cabeza);
    }
}

bool ListaCircular::eliminar(int id) {
    if (!cabeza) return false;

    // Caso especial: primer nodo
    if (cabeza->getDato().getId() == id) {
        if (cabeza->getSiguiente() == cabeza) {
            // Solo un nodo
            delete cabeza;
            cabeza = nullptr;
            return true;
        }

        // Más de un nodo
        Nodo* ultimoNodo = cabeza;
        while (ultimoNodo->getSiguiente() != cabeza) {
            ultimoNodo = ultimoNodo->getSiguiente();
        }

        Nodo* temp = cabeza;
        cabeza = cabeza->getSiguiente();
        ultimoNodo->setSiguiente(cabeza);
        delete temp;
        return true;
    }

    // Buscar en el resto de la lista
    Nodo* actual = cabeza;
    do {
        if (actual->getSiguiente()->getDato().getId() == id) {
            Nodo* aEliminar = actual->getSiguiente();
            actual->setSiguiente(aEliminar->getSiguiente());
            delete aEliminar;
            return true;
        }
        actual = actual->getSiguiente();
    } while (actual != cabeza);

    return false;
}

Persona* ListaCircular::buscar(int id) {
    if (!cabeza) return nullptr;

    Nodo* actual = cabeza;
    do {
        if (actual->getDato().getId() == id) {
            return &(actual->getDato());
        }
        actual = actual->getSiguiente();
    } while (actual != cabeza);

    return nullptr;
}

void ListaCircular::mostrar() {
    if (!cabeza) {
        std::cout << "Lista vacía.\n";
        return;
    }

    Nodo* actual = cabeza;
    do {
        actual->getDato().mostrar();
        actual = actual->getSiguiente();
    } while (actual != cabeza);
}

void ListaCircular::ordenar() {
    if (!cabeza || cabeza->getSiguiente() == cabeza) return;
    
    // Encontrar el último nodo
    Nodo* ultimo = cabeza;
    while (ultimo->getSiguiente() != cabeza) {
        ultimo = ultimo->getSiguiente();
    }
    
    quickSortUtil(cabeza, ultimo);
}

File: TrabajosDanny/ProgramasDePython/Listas_Python/ListaCircular.py

class NodoCircular:
    def __init__(self, dato):
        self.dato = dato
        self.siguiente = None

class ListaCircular:
    def __init__(self):
        self.cabeza = None

    def crear(self, dato):
        nuevo_nodo = NodoCircular(dato)
        if not self.cabeza:
            self.cabeza = nuevo_nodo
            self.cabeza.siguiente = self.cabeza
        else:
            actual = self.cabeza
            while actual.siguiente != self.cabeza:
                actual = actual.siguiente
            actual.siguiente = nuevo_nodo
            nuevo_nodo.siguiente = self.cabeza

    def buscar(self, dato):
        if not self.cabeza:
            return False
        actual = self.cabeza
        while True:
            if actual.dato == dato:
                return True
            actual = actual.siguiente
            if actual == self.cabeza:
                break
        return False

    def eliminar(self, dato):
        if not self.cabeza:
            return False
        if self.cabeza.dato == dato:
            if self.cabeza.siguiente == self.cabeza:
                self.cabeza = None
            else:
                actual = self.cabeza
                while actual.siguiente != self.cabeza:
                    actual = actual.siguiente
                actual.siguiente = self.cabeza.siguiente
                self.cabeza = self.cabeza.siguiente
            return True
        actual = self.cabeza
        while actual.siguiente != self.cabeza:
            if actual.siguiente.dato == dato:
                actual.siguiente = actual.siguiente.siguiente
                return True
            actual = actual.siguiente
        return False

    def mostrar(self):
        elementos = []
        if not self.cabeza:
            return elementos
        actual = self.cabeza
        while True:
            elementos.append(actual.dato)
            actual = actual.siguiente
            if actual == self.cabeza:
                break
        return elementos
    
    def buscar_por_cedula(self, cedula):
        persona = []
        actual = self.cabeza
        while actual:
            if actual.dato==cedula:
                persona.append(actual.dato)
                actual = actual.siguiente
                persona.append(actual.dato)
                actual = actual.siguiente
                persona.append(actual.dato)
                return persona
            else:
                actual = actual.siguiente
                actual = actual.siguiente
                if actual.siguiente:
                    actual = actual.siguiente
                else:
                    return None
                
    def buscar_por_nombres(self, nombre, apellido):
        persona = []
        actual = self.cabeza
        aux = self.cabeza
        while actual:
            aux = actual
            actual = actual.siguiente

            if actual.dato == nombre:
                persona.append(aux.dato)
                persona.append(actual.dato)
                actual = actual.siguiente
                if actual.dato == apellido:
                    persona.append(actual.dato)
                    return persona
            else:
                actual = actual.siguiente
                if actual.siguiente:
                    actual = actual.siguiente
                else:
                    return None

